{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cf28f912",
   "metadata": {},
   "source": [
    "## Task 1: generate simulated data\n",
    "\n",
    "We are going to generate sequences with motif: ATATTCA and GTACTGC.\n",
    "\n",
    "They will either appear together with a distance of 10-fold from each other: 10, 20, 30, .... nucleotides apart.\n",
    "\n",
    "Or only one of them will appear.\n",
    "\n",
    "And all the other part of the sequences will be randomly generated.\n",
    "\n",
    "And then we will add \"-\" tokens for the deletion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a33cf9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from dataclasses import dataclass\n",
    "from typing import List, Optional, Tuple\n",
    "\n",
    "DNA = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "MOTIF_A = \"ATATTCA\"\n",
    "MOTIF_B = \"GTACTGC\"\n",
    "\n",
    "\n",
    "def rand_dna(n: int, rng: random.Random) -> str:\n",
    "    return \"\".join(rng.choice(DNA) for _ in range(n))\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class Sequence:\n",
    "    seq: str\n",
    "    label: str               # \"both\", \"A_only\", \"B_only\", \"no_motif\"\n",
    "    deletions: int    # number of deletions in the sequence\n",
    "    pos_a: Optional[int]     # start index or None\n",
    "    pos_b: Optional[int]     # start index or None\n",
    "    gap: Optional[int]       # nt between motifs if both else None\n",
    "\n",
    "\n",
    "def make_example(\n",
    "    length: int,\n",
    "    mode: str,                     # \"both\" | \"A_only\" | \"B_only\" | \"no_motif\" \n",
    "    gaps: List[int],               # e.g. [10,20,30,...]\n",
    "    deletions: int,\n",
    "    rng: random.Random\n",
    ") -> Sequence:\n",
    "    if length < max(len(MOTIF_A), len(MOTIF_B)) + 1:\n",
    "        raise ValueError(\"Sequence length too short for motifs.\")\n",
    "    \n",
    "    # we don't allow gap not divisible by 10\n",
    "    invalid = [g for g in gaps if g % 10 != 0]\n",
    "    if invalid:\n",
    "        raise ValueError(\n",
    "            f\"Invalid gaps detected: {invalid}. \"\n",
    "            \"All gaps must be divisible by 10 (e.g. 10, 20, 30, ...).\"\n",
    "        )\n",
    "    \n",
    "    if mode == \"both\":\n",
    "        gap = rng.choice(gaps)\n",
    "        total_motif_len = len(MOTIF_A) + gap + len(MOTIF_B)\n",
    "        if total_motif_len > length:\n",
    "            raise ValueError(\n",
    "                f\"Length {length} too short for both motifs with gap {gap} \"\n",
    "                f\"(need >= {total_motif_len}).\"\n",
    "            )\n",
    "\n",
    "\n",
    "        start = rng.randint(0, length - total_motif_len)\n",
    "        prefix_len = start\n",
    "        suffix_len = length - prefix_len - total_motif_len\n",
    "\n",
    "        prefix = rand_dna(prefix_len, rng)\n",
    "        between = rand_dna(gap, rng)\n",
    "        suffix = rand_dna(suffix_len, rng)\n",
    "\n",
    "\n",
    "        seq = prefix + MOTIF_A + between + MOTIF_B + suffix\n",
    "        pos_a = prefix_len\n",
    "        pos_b = prefix_len + len(MOTIF_A) + gap\n",
    "\n",
    "\n",
    "        return Sequence(seq=seq, label=\"both\", deletions = deletions, pos_a=pos_a, pos_b=pos_b, gap=gap)\n",
    "\n",
    "    elif mode == \"A_only\":\n",
    "        total = len(MOTIF_A)\n",
    "        start = rng.randint(0, length - total)\n",
    "        seq = rand_dna(start, rng) + MOTIF_A + rand_dna(length - start - total, rng)\n",
    "        return Sequence(seq=seq, label=\"A_only\", deletions = deletions, pos_a=start, pos_b=None, gap=None)\n",
    "\n",
    "    elif mode == \"B_only\":\n",
    "        total = len(MOTIF_B)\n",
    "        start = rng.randint(0, length - total)\n",
    "        seq = rand_dna(start, rng) + MOTIF_B + rand_dna(length - start - total, rng)\n",
    "        return Sequence(seq=seq, label=\"B_only\", deletions = deletions, pos_a=None, pos_b=start, gap=None)\n",
    "\n",
    "    elif mode == \"no_motif\":\n",
    "        seq = rand_dna(length, rng)\n",
    "        return Sequence(seq=seq, label=\"no_motif\", deletions = deletions, pos_a=None, pos_b=None, gap=None)\n",
    "    \n",
    "    else:\n",
    "        raise ValueError(\"mode must be: 'both', 'A_only', or 'B_only'\")\n",
    "\n",
    "\n",
    "def generate_dataset(\n",
    "    n: int, # numbers of seuqneces we want to generate\n",
    "    length: int = 120, # lenght of each sequence, should all be the same\n",
    "    gaps: List[int] = None, # list of possible gaps between motif A and motif B: 10, 20, 30, ..., 100\n",
    "    deletions: int = 0, # number of deletions to augment each sequence with\n",
    "    p_both: float = 0.4, # how many sequences contain both motifs in prbobability\n",
    "    p_a_only: float = 0.25,  # how many sequences contain only motif A in probability\n",
    "    p_b_only: float = 0.25,  # how many sequences contain only motif B in probability\n",
    "    p_no_motif: float = 0.1, # how many sequences contain no motif in probability\n",
    "    seed: int = 727 # random seed\n",
    ") -> List[Sequence]:\n",
    "    \n",
    "    if gaps is None:\n",
    "        gaps = list(range(10, 101, 10))  # 10,20,...,100\n",
    "\n",
    "    if abs((p_both + p_a_only + p_b_only + p_no_motif) - 1.0) > 1e-9:\n",
    "        raise ValueError(\"Probabilities must sum to 1 (p_both + p_a_only + p_b_only).\")\n",
    "\n",
    "    rng = random.Random(seed)\n",
    "\n",
    "    data = []\n",
    "    for _ in range(n):\n",
    "        r = rng.random()\n",
    "        # r is in range [0,1), so we can use it to randomly select mode, given the probabilities\n",
    "        if r < p_both:\n",
    "            mode = \"both\"\n",
    "        elif r < p_both + p_a_only:\n",
    "            mode = \"A_only\"\n",
    "        elif r < p_both + p_a_only + p_b_only:\n",
    "            mode = \"B_only\"\n",
    "        else:\n",
    "            mode = \"no_motif\"\n",
    "\n",
    "        ex = make_example(length=length, mode=mode, gaps=gaps, deletions = deletions, rng=rng)\n",
    "        data.append(ex)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "n=100\n",
    "length=120\n",
    "dataset = generate_dataset(n=n, length=length, deletions=0)\n",
    "\n",
    "# output the sequences in FASTA format\n",
    "def write_fasta(dataset, filepath):\n",
    "    with open(filepath, \"w\") as f:\n",
    "        for i, ex in enumerate(dataset, start=1):\n",
    "            header = (\n",
    "                f\">seq{i:04d}\"\n",
    "                f\"|label={ex.label}\"\n",
    "                f\"|posAmotif={ex.pos_a}\"\n",
    "                f\"|posBmotif={ex.pos_b}\"\n",
    "                f\"|gaplength={ex.gap}\"\n",
    "                f\"|deletions={ex.deletions}\"\n",
    "            )\n",
    "            f.write(header + \"\\n\")\n",
    "            f.write(ex.seq + \"\\n\")\n",
    "            \n",
    "\n",
    "write_fasta(dataset, f\"simulated_sequences/sequence_size{n}_length{length}.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c23b38f",
   "metadata": {},
   "source": [
    "## Data augumentation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0fdc6ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# For data augmentation, a ‘-’ token can be added anywhere in the generated sequences.\n",
    "\n",
    "# number of deletions controll how many deletions we want per sequence\n",
    "def augment_with_deletions(dataset: List[Sequence], max_number_of_dels: int = 20, no_del_seqs = 0.25, seed: int = 727) -> List[Sequence]:\n",
    "    rng = random.Random(seed)\n",
    "    augmented_data = []\n",
    "    \n",
    "    size_seq_dels = int((1 - no_del_seqs) * len(dataset))\n",
    "    dataset_with_deletions = random.sample(dataset, size_seq_dels)\n",
    "    \n",
    "    for ex in dataset:\n",
    "        # augmented_data.append(ex)  # original sequence should be included or not?\n",
    "        if ex in dataset_with_deletions:\n",
    "            number_of_dels = rng.randint(1, max_number_of_dels)\n",
    "            for _ in range(number_of_dels):\n",
    "                pos = rng.randint(0, len(ex.seq) - 1)  # position to insert gap\n",
    "                ex.seq = ex.seq[:pos] + '-' + ex.seq[pos + 1:] \n",
    "            augmented_ex = Sequence(\n",
    "                    seq=ex.seq,\n",
    "                    label=ex.label,\n",
    "                    deletions=number_of_dels,\n",
    "                    pos_a=ex.pos_a,\n",
    "                    pos_b=ex.pos_b,\n",
    "                    gap=ex.gap\n",
    "                )\n",
    "            augmented_data.append(augmented_ex)\n",
    "        else:\n",
    "            augmented_data.append(ex)   \n",
    "\n",
    "    return augmented_data\n",
    "\n",
    "max_number_of_dels = 20\n",
    "no_del_seqs = 0.25\n",
    "augumented_dataset = augment_with_deletions(dataset, max_number_of_dels=max_number_of_dels, no_del_seqs=no_del_seqs)\n",
    "write_fasta(augumented_dataset, f\"simulated_sequences/augumented_sequence_size{n}_length{length}_deletions{max_number_of_dels}_nodeletionseq{no_del_seqs}.fasta\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sysgen_project",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.25"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
