{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cf28f912",
   "metadata": {},
   "source": [
    "## Task 1: generate simulated data with deletions\n",
    "\n",
    "We are going to generate sequences with motif: ATATTCA and GTACTGC.\n",
    "\n",
    "They will either appear together with a nulcleotid distance of 10-fold from each other: 10, 20, 30, .... nucleotides apart. (where the deletion don't count), but there can be deletions inside.\n",
    "\n",
    "Or only one of them will appear.\n",
    "\n",
    "And all the other part of the sequences will be randomly generated with deletions.\n",
    "\n",
    "And then we will add \"-\" tokens for the deletion. Should we have deletion in the motif too? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a33cf9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from dataclasses import dataclass\n",
    "from typing import List, Optional, Tuple\n",
    "\n",
    "DNA = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "MOTIF_A = \"ATATTCA\"\n",
    "MOTIF_B = \"GTACTGC\"\n",
    "\n",
    "\n",
    "def rand_dna(n: int, rng: random.Random) -> str:\n",
    "    return \"\".join(rng.choice(DNA) for _ in range(n))\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class Sequence:\n",
    "    seq: str\n",
    "    label: str               # \"both\", \"A_only\", \"B_only\", \"no_motif\"\n",
    "    deletions: int    # numbers of deletions in the sequence\n",
    "    pos_a: Optional[int]     # start index or None\n",
    "    pos_b: Optional[int]     # start index or None\n",
    "    gap: Optional[int]       # nt between motifs if both else None\n",
    "\n",
    "import math\n",
    "\n",
    "# this function can be used to augment the random background with deletions\n",
    "def augmentation(seq: str, frac: float, rng: random.Random) -> str:\n",
    "    if frac <= 0:\n",
    "        return seq\n",
    "    n = len(seq)\n",
    "    k = max(1, math.floor(frac * n))\n",
    "    positions = list(range(n))\n",
    "    rng.shuffle(positions)\n",
    "    replace_pos = positions[:k]\n",
    "    s = list(seq)\n",
    "    for p in replace_pos:\n",
    "        s[p] = '-'\n",
    "    return \"\".join(s)\n",
    "\n",
    "def make_example(\n",
    "    length: int,\n",
    "    mode: str,                     # \"both\" | \"A_only\" | \"B_only\" | \"no_motif\" \n",
    "    gaps: List[int],               # e.g. [10,20,30,...]\n",
    "    deletions: float,\n",
    "    if_deletions: bool,\n",
    "    rng: random.Random\n",
    ") -> Sequence:\n",
    "    if length < max(len(MOTIF_A), len(MOTIF_B)) + 1:\n",
    "        raise ValueError(\"Sequence length too short for motifs.\")\n",
    "    \n",
    "    # we don't allow gap not divisible by 10\n",
    "    invalid = [g for g in gaps if g % 10 != 0]\n",
    "    if invalid:\n",
    "        raise ValueError(\n",
    "            f\"Invalid gaps detected: {invalid}. \"\n",
    "            \"All gaps must be divisible by 10 (e.g. 10, 20, 30, ...).\"\n",
    "        )\n",
    "    \n",
    "    if mode == \"both\":\n",
    "        gap = rng.choice(gaps)\n",
    "        total_motif_len = len(MOTIF_A) + gap + len(MOTIF_B)\n",
    "        if total_motif_len > length:\n",
    "            raise ValueError(\n",
    "                f\"Length {length} too short for both motifs with gap {gap} \"\n",
    "                f\"(need >= {total_motif_len}).\"\n",
    "            )\n",
    "\n",
    "        if if_deletions:\n",
    "            deletions_in_gap = rng.randint(1, int(gap * deletions))  # up to 10% deletions in the gap\n",
    "        else:\n",
    "            deletions_in_gap = 0\n",
    "        \n",
    "        start = rng.randint(0, length - total_motif_len - deletions_in_gap)\n",
    "        prefix_len = start\n",
    "        suffix_len = length - prefix_len - total_motif_len - deletions_in_gap\n",
    "\n",
    "        # generate random background with deletions, the deletion in the gap will be added laer\n",
    "        if if_deletions:\n",
    "            prefix = augmentation(rand_dna(prefix_len, rng), deletions, rng)\n",
    "            between = rand_dna(gap, rng)\n",
    "            suffix = augmentation(rand_dna(suffix_len, rng), deletions, rng)\n",
    "        else:\n",
    "            prefix = rand_dna(prefix_len, rng)\n",
    "            between = rand_dna(gap, rng)\n",
    "            suffix = rand_dna(suffix_len, rng)\n",
    "        \n",
    "        # motifs region -> by adding them, so the nucleotide length remains the same\n",
    "        if deletions_in_gap > 0:\n",
    "            indices = list(range(gap))\n",
    "            rng.shuffle(indices)\n",
    "            del_indices = set(indices[:deletions_in_gap])\n",
    "            between = \"\".join(\n",
    "                \"-\" + c if i in del_indices else c\n",
    "                for i, c in enumerate(between)\n",
    "            )\n",
    "\n",
    "        seq = prefix + MOTIF_A + between + MOTIF_B + suffix\n",
    "        pos_a = prefix_len\n",
    "        pos_b = prefix_len + len(MOTIF_A) + gap + deletions_in_gap\n",
    "        deletions = seq.count('-')\n",
    "\n",
    "        return Sequence(seq=seq, label=\"both\", deletions = deletions, pos_a=pos_a, pos_b=pos_b, gap=gap)\n",
    "\n",
    "    elif mode == \"A_only\":\n",
    "        total = len(MOTIF_A)\n",
    "        start = rng.randint(0, length - total)\n",
    "        if if_deletions:\n",
    "            seq = augmentation(rand_dna(start, rng), deletions, rng) + MOTIF_A + augmentation(rand_dna(length - start - total, rng), deletions, rng)\n",
    "        else:\n",
    "            seq = rand_dna(start, rng) + MOTIF_A + rand_dna(length - start - total, rng)\n",
    "        deletions = seq.count('-')\n",
    "        return Sequence(seq=seq, label=\"A_only\", deletions = deletions, pos_a=start, pos_b=None, gap=None)\n",
    "\n",
    "    elif mode == \"B_only\":\n",
    "        total = len(MOTIF_B)\n",
    "        start = rng.randint(0, length - total)\n",
    "        if if_deletions:\n",
    "            seq = augmentation(rand_dna(start, rng), deletions, rng) + MOTIF_B + augmentation(rand_dna(length - start - total, rng), deletions, rng)\n",
    "        else:\n",
    "            seq = rand_dna(start, rng) + MOTIF_B + rand_dna(length - start - total, rng)\n",
    "        deletions = seq.count('-')\n",
    "        return Sequence(seq=seq, label=\"B_only\", deletions = deletions, pos_a=None, pos_b=start, gap=None)\n",
    "\n",
    "    elif mode == \"no_motif\":\n",
    "        if if_deletions:\n",
    "            seq = augmentation(rand_dna(length, rng), deletions, rng)\n",
    "        else:\n",
    "            seq = rand_dna(length, rng)\n",
    "        deletions = seq.count('-')\n",
    "        return Sequence(seq=seq, label=\"no_motif\", deletions = deletions, pos_a=None, pos_b=None, gap=None)\n",
    "    \n",
    "    else:\n",
    "        raise ValueError(\"mode must be: 'both', 'A_only', or 'B_only'\")\n",
    "\n",
    "\n",
    "def generate_dataset(\n",
    "    n: int, # numbers of seuqneces we want to generate\n",
    "    length: int = 120, # lenght of each sequence, should all be the same\n",
    "    gaps: List[int] = None, # list of possible gaps between motif A and motif B: 10, 20, 30, ..., 100\n",
    "    deletions: float = 0.1, # percentage of deletions to augment each sequence with\n",
    "    no_del_seqs: float = 0.25, # percentage of sequences without any deletions\n",
    "    p_both: float = 0.4, # how many sequences contain both motifs in prbobability\n",
    "    p_a_only: float = 0.25,  # how many sequences contain only motif A in probability\n",
    "    p_b_only: float = 0.25,  # how many sequences contain only motif B in probability\n",
    "    p_no_motif: float = 0.1, # how many sequences contain no motif in probability\n",
    "    seed: int = 727 # random seed\n",
    ") -> List[Sequence]:\n",
    "    \n",
    "    if gaps is None:\n",
    "        gaps = list(range(10, 101, 10))  # 10,20,...,100\n",
    "\n",
    "    if abs((p_both + p_a_only + p_b_only + p_no_motif) - 1.0) > 1e-9:\n",
    "        raise ValueError(\"Probabilities must sum to 1 (p_both + p_a_only + p_b_only).\")\n",
    "\n",
    "    rng = random.Random(seed)\n",
    "\n",
    "    data = []\n",
    "    for _ in range(n):\n",
    "        r = rng.random()\n",
    "        # r is in range [0,1), so we can use it to randomly select mode, given the probabilities\n",
    "        if r < p_both:\n",
    "            mode = \"both\"\n",
    "        elif r < p_both + p_a_only:\n",
    "            mode = \"A_only\"\n",
    "        elif r < p_both + p_a_only + p_b_only:\n",
    "            mode = \"B_only\"\n",
    "        else:\n",
    "            mode = \"no_motif\"\n",
    "\n",
    "        if rng.random() < no_del_seqs:\n",
    "            if_deletions = False\n",
    "        else:\n",
    "            if_deletions = True\n",
    "\n",
    "        ex = make_example(length=length, mode=mode, gaps=gaps, deletions = deletions, if_deletions = if_deletions, rng=rng)\n",
    "        data.append(ex)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "n=1000\n",
    "length=150\n",
    "deletions = 0.1\n",
    "no_del_seqs = 0.25\n",
    "gaps = list(range(10, 101, 10))\n",
    "\n",
    "dataset = generate_dataset(n=n, length=length, gaps = gaps, deletions=deletions, no_del_seqs=no_del_seqs, seed=)\n",
    "\n",
    "# output the sequences in FASTA format\n",
    "def write_fasta(dataset, filepath):\n",
    "    with open(filepath, \"w\") as f:\n",
    "        for i, ex in enumerate(dataset, start=1):\n",
    "            header = (\n",
    "                f\">seq{i:04d}\"\n",
    "                f\"|label={ex.label}\"\n",
    "                f\"|posAmotif={ex.pos_a}\"\n",
    "                f\"|posBmotif={ex.pos_b}\"\n",
    "                f\"|gaplength={ex.gap}\"\n",
    "                f\"|deletions={ex.deletions}\"\n",
    "            )\n",
    "            f.write(header + \"\\n\")\n",
    "            f.write(ex.seq + \"\\n\")\n",
    "            \n",
    "\n",
    "write_fasta(dataset, f\"test_data/augumented_sequence_size{n}_length{length}_deletions{deletions}_nodeletionseq{no_del_seqs}.fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p06_indel_GLM_W2526",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
