{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Training and Running of new GLM model\n",
    "This Notebook is for training and running the our new GLM model that includes (for now only) deletion tokens"
   ],
   "id": "11397fdeb6e87ca4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Adding necessary imports\n",
    "You can run this block to import necessary classes and functions"
   ],
   "id": "920c2ce5af494e42"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from fundemental_classes.model_related.glm_model import GLMModel\n",
    "\n",
    "# You can change force_retrain to True if you want to always retrain your model everytime\n",
    "deletion_model = GLMModel(\"./dna_bert_final_indel\",\n",
    "                          \"simulated_sequences_distance=5/new_augumented_sequence_size10000_length150_deletions0.2_nodeletionseq0.05.fasta\",\n",
    "                          force_retrain=False, include_deletions=True)\n",
    "baseline_model = GLMModel(\"./dna_bert_final_baseline\",\n",
    "                           \"simulated_baseline_sequences_distance=5/new_augumented_sequence_size10000_length150_deletions0_nodeletionseq0.05.fasta\",\n",
    "                           force_retrain=False, include_deletions=False)\n",
    "\n",
    "# Set the default Model to train and use for visualization\n",
    "model = baseline_model\n",
    "use_deletions = False\n"
   ],
   "id": "48824261cb150db9"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Training\n",
    "We use `Bert` model to create the embeddings and train it using masking to get nicely trained model. If you want to change the size of training data, please look into `simulated_sequences` directory. \\\\\\\n",
    "\n",
    "Below is given two subheaders: *Baseline Model Training & Deletion Model Training*. Each of them can be run seperately if you are looking for seperate results or loading.\n",
    "*important:* Training is not necessary if you already have a trained model. You can load it using the `GLMModel`."
   ],
   "id": "ea084542235425f3"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Baseline Model Training",
   "id": "ab9eb5608cfe269d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "baseline_model.train(epochs=500, batch_size=32, lr=2e-4)",
   "id": "9055c3fe9ea5e52c"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### Deletion Model Training",
   "id": "754c57dedd8babfa"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "deletion_model.train(epochs=500, batch_size=32, lr=2e-4)\n",
   "id": "5aa5f340a31254be"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Visualization | Sequence reconstruction probabilities\n",
    "This block is used for getting already trained model, passing the sequence to test how well the model performs.\n",
    "\\\\\\\\\\\n",
    "*Important:* For now we pass our test cases by hand. However in the future we will generate good inputs to test how well the model functions"
   ],
   "id": "e670e1e248779cd7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### OLD SEQUENCE PLOTTER",
   "id": "a08e9677f9471ba2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from fundemental_classes.visualization.sequence_plotter import plot\n",
    "\n",
    "#normal reconstruction\n",
    "header = \">seq0003|label=both|posAmotif=5,98|posBmotif=17|gaplength=5|deletions=0\"\n",
    "sequence = \"GGTCGATATTCATTCCTGTACTGCTTGATCCACTACCATGTACTTCAATCCTGGCGATGTCTAGTCTCACAGGGAGGTACATATGGTGCAACATTGGAATATTCAACAGGGTATAGGCCTAGACGAGGCATATAAAGCCTGACAACATGT\"\n",
    "\n",
    "prob_matrix = model.get_full_reconstruction_probs(sequence)\n",
    "\n",
    "plot(header, sequence, prob_matrix, motif_length=6, with_deletions=use_deletions)"
   ],
   "id": "e2af24205b155704"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "### NEW SEQUENCE PLOTTER",
   "id": "e991ac9a63ec16bc"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from fundemental_classes.visualization.sequence_plotter_new import plot\n",
    "\n",
    "#normal reconstruction\n",
    "header = \"seq0003|label=both|posAmotif=15,58|posBmotif=28|gaplength=5|deletions=6\"\n",
    "sequence = \"GG-CCATCTTAGACCATATTCA-CCGCCGTACTGCATGAAG-CTGACGGAGTATGACAATATTCATCACCAGGACACGATCACGCTA-GTACTCAGCGGGCT-CTTCCTTGG-AGGAAGAGGGCCACTACCGGAACCGTTATTAATACGGATTCTT\"\n",
    "\n",
    "prob_matrix = model.get_full_reconstruction_probs(sequence)\n",
    "\n",
    "plot(header, sequence, prob_matrix, motif_length=6, with_deletions=use_deletions)"
   ],
   "id": "8387832f108540dd"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#replace with letter (SNP variation)\n",
    "header = \"seq0030|label=both|posAmotif=13|posBmotif=140|gaplength=100|deletions=27\"\n",
    "sequence = \"A-TTGGAGC-CCAATATTCA--CGCCTC-A-CC-A-ACTCG-AGGTTC--GGAAT-TCCCC-T-TCCCAGCCTTG-AGTC--AATTGGGG-TCCTGCTGA-TACAATGG--CT-GCTGTGGACGACTCCTATGATGGGCTGTACTGC--GA-G-GCTTCGGTAGAAT-CGTTATGCC\"\n",
    "\n",
    "prob_matrix = model.get_full_reconstruction_probs(sequence)\n",
    "\n",
    "plot(header, sequence, prob_matrix, motif_length=6, with_deletions=use_deletions)"
   ],
   "id": "1509f3811f9ef7fb"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#replace with deletion token (InDel variation)\n",
    "header = \"seq0030|label=both|posAmotif=13|posBmotif=140|gaplength=100|deletions=27\"\n",
    "sequence = \"A-TTGGAGC-CCAATATTCA--CGCCTC-A-CC-A-ACTCG-AGGTTC--GGAAT-TCCCC-T-TCCCA-CCTTG-AGTC--AATTGGGG-TCCTGCTGA-TACAATGG--CT-GCTGTGGACGACTCCTATGATGGGCTGTACTGC--GA-G-GCTTCGGTAGAAT-CGTTATGCC\"\n",
    "\n",
    "prob_matrix = model.get_full_reconstruction_probs(sequence)\n",
    "\n",
    "plot(header, sequence, prob_matrix, motif_length=6, with_deletions=use_deletions)"
   ],
   "id": "57d6ce2dcbc44841"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Visualization | Dependency Map\n",
    "This block is used for getting already trained model, passing the sequence to test how well the model performs.\n",
    "\\\\\\\\\\"
   ],
   "id": "da55dd8d22c6d46d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from fundemental_classes.visualization.heat_map_plotter import DependencyMapGenerator\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "from fundemental_classes.visualization.sequence_logo import SequenceLogo\n",
    "\n",
    "\n",
    "# a function for creating the heatmap with overlayed logo with input sequence, type of dependency and model\n",
    "def depencency_map_with_logo(model, sequence, type='snp'):\n",
    "    dep_gen = DependencyMapGenerator(model, type=type, use_deletions=use_deletions)\n",
    "    prob_matrix = model.get_full_reconstruction_probs(sequence)\n",
    "    dep_map = dep_gen.analyze(sequence, show_plot=False)\n",
    "    L = dep_map.shape[0]\n",
    "    reconstruction_logo = SequenceLogo.from_reconstruction(prob_matrix)\n",
    "    fig = go.Figure()\n",
    "    fig.add_trace(\n",
    "        go.Heatmap(\n",
    "            z=dep_map,\n",
    "            x=np.arange(L),\n",
    "            y=np.arange(L),\n",
    "            colorscale=\"RdBu_r\",\n",
    "        )\n",
    "    )\n",
    "    # if the xaxes and yaxes are shown\n",
    "    fig.update_xaxes(visible=True, scaleanchor=\"y\", constrain=\"domain\")\n",
    "    fig.update_yaxes(visible=True, autorange=\"reversed\", scaleanchor=\"x\", constrain=\"domain\")\n",
    "    # Overlay logo at the top \n",
    "    fig.add_layout_image(\n",
    "        source=reconstruction_logo.to_svg(data_url=True),\n",
    "        xref=\"x\",\n",
    "        yref=\"y domain\",\n",
    "        x=-0.5,  # align left edge of first column\n",
    "        y=1.0,  # top of heatmap domain\n",
    "        sizex=L,  # span exactly L columns\n",
    "        sizey=0.10,  # height in domain units (tune)\n",
    "        xanchor=\"left\",\n",
    "        yanchor=\"bottom\",\n",
    "        sizing=\"stretch\",\n",
    "    )\n",
    "    fig.update_layout(\n",
    "        width=900,\n",
    "        height=900,\n",
    "        margin=dict(l=20, r=20, t=100, b=20),\n",
    "    )\n",
    "    fig.show()\n"
   ],
   "id": "1dcf218a166670a9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#dependency map for snp \n",
    "header = \"seq0004|label=both|posAmotif=22,44,85|posBmotif=35,57|gaplength=5|deletions=9\"\n",
    "sequence = \"CAGTTTA-TTCTTACCCCCCTAATATTCACGCG-GGTACTGC-CATATTCACCA-GCGTACTGCTTTAC-GGAGACTA-TCTGTAATATTCATAATACGGCTATAACAACAATTTGATGGC-AGCAGGAGAGCAGACCAAAAGTG-GG-TGTTAAGTCA\"\n",
    "dep_map = depencency_map_with_logo(model, sequence, type='snp')"
   ],
   "id": "1d53fb3ddc23fda7"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#dependency map for indels - reconstruction\n",
    "\n",
    "header = \"seq0004|label=both|posAmotif=22,44,85|posBmotif=35,57|gaplength=5|deletions=9\"\n",
    "sequence = \"CAGTTTA-TTCTTACCCCCCTAATATTCACGCG-GGTACTGC-CATATTCACCA-GCGTACTGCTTTAC-GGAGACTA-TCTGTAATATTCATAATACGGCTATAACAACAATTTGATGGC-AGCAGGAGAGCAGACCAAAAGTG-GG-TGTTAAGTCA\"\n",
    "\n",
    "dep_map = depencency_map_with_logo(model, sequence, type='indel')"
   ],
   "id": "7645272aede806ff"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#dependency map for indels - replacement\n",
    "\n",
    "header = \"seq0030|label=both|posAmotif=13|posBmotif=140|gaplength=100|deletions=27\"\n",
    "sequence = \"A-TTGGAGC-CCAATATTCA--CGCCTC-A-CC-A-ACTCG-AGGTTC--GGAAT-TCCCC-T-TCCCA---TTG-AGTC--AATTGGGG-TCCTGCTGA-TACAATGG--CT-GCTGTGGACGACTCCTATGATGGGCTGTACTGC--GA-G-GCTTCGGTAGAAT-CGTTATGCC\"\n",
    "\n",
    "dep_map = depencency_map_with_logo(model, sequence, type='indel')"
   ],
   "id": "5b9093c74845154a"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "#dependency map for indels - removal\n",
    "\n",
    "header = \"seq0004|label=both|posAmotif=22,44,85|posBmotif=35,57|gaplength=5|deletions=9\"\n",
    "sequence = \"CAGTTTA-TTCTTACCCCCCTAATATTCACGCG-GGTACTGC-CATATTCACCA-GCGTACTGCTTTAC-GGAGACTA-TCTGTAATATTCATAATACGGCTATAACAACAATTTGATGGC-AGCAGGAGAGCAGACCAAAAGTG-GG-TGTTAAGTCA\"\n",
    "\n",
    "dep_map = depencency_map_with_logo(model, sequence, type='indel')"
   ],
   "id": "fb6db159d62b2e06"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p06_indel_GLM_W2526",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
